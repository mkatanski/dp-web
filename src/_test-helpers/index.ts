export * from "./withResponsiveContext";
export * from "./withThemeProvider";
