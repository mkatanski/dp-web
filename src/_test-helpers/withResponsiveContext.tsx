import React from "react";
import { Context as ResponsiveContext } from "react-responsive";
import { Breakpoints } from "config/breakpoints";

export const withMobileSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.mobile }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withTabletSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.tablet }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.desktop }}>
    {Component}
  </ResponsiveContext.Provider>
);

export const withLargeDesktopSize = (Component: React.ReactNode) => (
  <ResponsiveContext.Provider value={{ width: Breakpoints.largeDesktop }}>
    {Component}
  </ResponsiveContext.Provider>
);
