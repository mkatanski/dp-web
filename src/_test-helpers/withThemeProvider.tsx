import React from "react";
import { ThemeProvider as DefaultThemeProvider } from "lib/theme/ThemeProvider";
import { StyledTheme, Theme } from "styles/theme/_types";
import { createMuiTheme } from "@material-ui/core";

const testTheme: StyledTheme = {
  primary: "primary",
  secondary: "secondary",
  textPrimary: "textPrimary",
  textSecondary: "textSecondary",
  textInverse: "textInverse"
};

const muiTheme = createMuiTheme({
  direction: "ltr"
});

const theme: Theme = Object.assign(testTheme, muiTheme);

export const withThemeProvider = (Component: React.ReactNode) => (
  <DefaultThemeProvider theme={theme}>{Component}</DefaultThemeProvider>
);
