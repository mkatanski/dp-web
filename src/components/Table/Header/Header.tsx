import React from "react";
import styled from "styled-components";
import { TableCell, TableCellProps, TableSortLabel } from "@material-ui/core";

export type HeaderProps = TableCellProps & {
  width?: number | "inherit" | "auto";
  showSortLabel?: boolean;
};

const StyledHeader = styled<React.ComponentType<HeaderProps>>(TableCell)`
  cursor: ${({ onClick }) => (onClick ? "pointer" : "default")};
  && {
    min-width: ${({ width = "inherit" }) =>
      typeof width === "number" ? `${width}px` : width};
  }
`;

export const Header: React.FC<HeaderProps> = ({
  children,
  sortDirection,
  showSortLabel,
  ...rest
}: HeaderProps) => (
  <StyledHeader {...rest}>
    {showSortLabel && (
      <TableSortLabel
        active={!!sortDirection}
        direction={sortDirection || undefined}
      >
        {children}
      </TableSortLabel>
    )}

    {!showSortLabel && children}
  </StyledHeader>
);
