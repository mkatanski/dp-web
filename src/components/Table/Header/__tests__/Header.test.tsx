import React from "react";
import TestRenderer from "react-test-renderer";
import "jest-styled-components";

/* Uncomment below if needed */
import { withThemeProvider } from "_test-helpers";
import { render, fireEvent } from "@testing-library/react";

import { Header } from "../Header";

/* Uncomment below if You need/want to use enzyme */
// import enzyme from "enzyme";

/* Component mock example */
// jest.mock("[module]", () => ({
//   [ComponentName]: ({children, ...rest}: React.PropsWithChildren<{}>) => (
//     <div id="[ComponentName]" {...rest}>{children}</div>
//   )
// }));

describe("components", () => {
  describe("Header", () => {
    it("should render Header component", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Header showSortLabel />)
      );

      expect(tree).toMatchSnapshot();
    });

    it("should render Header component with specific width", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Header width={145} showSortLabel />)
      );

      expect(tree).toMatchSnapshot();
    });

    it("should render Header component with ASC sort order", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Header sortDirection="asc" showSortLabel />)
      );

      expect(tree).toMatchSnapshot();
    });

    it("should render Header component with DESC sort order", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Header sortDirection="desc" showSortLabel />)
      );

      expect(tree).toMatchSnapshot();
    });

    it("should render Header component with onClick", () => {
      const tree = TestRenderer.create(
        withThemeProvider(<Header onClick={() => null} showSortLabel />)
      );

      expect(tree).toMatchSnapshot();
    });

    it("should handle onClick event", async () => {
      const handleClick = jest.fn();
      const tableRow = document.createElement("tr");
      const { findByText } = render(
        withThemeProvider(<Header onClick={handleClick}>Test</Header>),
        {
          container: document.body.appendChild(tableRow)
        }
      );
      fireEvent.click(await findByText("Test"));

      expect(handleClick).toHaveBeenCalledTimes(1);
    });
  });
});
