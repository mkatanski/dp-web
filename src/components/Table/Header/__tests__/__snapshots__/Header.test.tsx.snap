// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components Header should render Header component 1`] = `
.c0 {
  cursor: default;
}

.c0.c0 {
  min-width: inherit;
}

<td
  className="c0"
  data-props={Object {}}
  id="MUI.TableCell"
>
  <div
    data-props={
      Object {
        "active": false,
        "direction": undefined,
      }
    }
    id="MUI.TableSortLabel"
  />
</td>
`;

exports[`components Header should render Header component with ASC sort order 1`] = `
.c0 {
  cursor: default;
}

.c0.c0 {
  min-width: inherit;
}

<td
  className="c0"
  data-props={Object {}}
  id="MUI.TableCell"
>
  <div
    data-props={
      Object {
        "active": true,
        "direction": "asc",
      }
    }
    id="MUI.TableSortLabel"
  />
</td>
`;

exports[`components Header should render Header component with DESC sort order 1`] = `
.c0 {
  cursor: default;
}

.c0.c0 {
  min-width: inherit;
}

<td
  className="c0"
  data-props={Object {}}
  id="MUI.TableCell"
>
  <div
    data-props={
      Object {
        "active": true,
        "direction": "desc",
      }
    }
    id="MUI.TableSortLabel"
  />
</td>
`;

exports[`components Header should render Header component with onClick 1`] = `
.c0 {
  cursor: pointer;
}

.c0.c0 {
  min-width: inherit;
}

<td
  className="c0"
  data-props={Object {}}
  id="MUI.TableCell"
  onClick={[Function]}
>
  <div
    data-props={
      Object {
        "active": false,
        "direction": undefined,
      }
    }
    id="MUI.TableSortLabel"
  />
</td>
`;

exports[`components Header should render Header component with specific width 1`] = `
.c0 {
  cursor: default;
}

.c0.c0 {
  min-width: 145px;
}

<td
  className="c0"
  data-props={
    Object {
      "width": 145,
    }
  }
  id="MUI.TableCell"
>
  <div
    data-props={
      Object {
        "active": false,
        "direction": undefined,
      }
    }
    id="MUI.TableSortLabel"
  />
</td>
`;
